enhancements/automated-service-ca-rotation.md

```diff
@@ -0,0 +1,271 @@
+---
+title: Automated Service CA Rotation
+authors:
+  - "@marun"
+reviewers:
+  - "@deads2k"
+  - "@mfojtik"
+  - "@mrogers950"
+  - "@stlaz"
+  - "@sttts"
+approvers:
+  - TBD
+creation-date: 2019-10-23
+last-updated: 2019-10-23
+status: implementable
+see-also:
+  - "https://docs.google.com/document/d/1VqrIERs30M9EyPaqcO43wcSXHXvfe_Ao2W0ThDIohlM/edit"
+replaces:
+superseded-by:
+---
+
+# automated-service-ca-rotation
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift/docs]
+
+## Summary
+
+Enable automated rotation of the service signing CA and ensure that clients
+(using an operator-provided CA bundle to verify trust) and servers (serving with
+an operator-provided certificate) will be able to continue communicating without
+undue disruption.
+
+## Motivation
+
+Service serving certificates are automatically renewed by the service CA
+controller. However, the service CA signer is not. The service CA is expired
+after 1 year, and once that expiration has passed clients will not be able to
+validate server certs, leading to service disruption. Recovering from this
+currently requires manual intervention (force rotation of the CA by deleting the
+CA secret).
+
+Since 4.1 shipped without automated rotation, operators depending on the service
+CA operator will break without manual triggering of forced rotation via deletion
+of the signing secret. Since manually-triggered rotation won't provide a grace
+period where the old and new CAs are trusted, it is likely to induce a brief
+period of system instability while servers and clients converge on the use of the
+new trust chain.
+
+### Goals
+
+- The signing CA should be automatically rotated before expiry
+
+- Rotation should prevent trust breakage between consumers of the operator as
+  long as the pre-rotation signing CA remains valid. The following should be
+  true:
+  - A serving cert, from the new CA, validates against the pre-rotation CA
+    bundle.
+  - A serving cert, from the old CA, validates against the post-rotation CA
+    bundle.
```

Comment:
by transitivity this implies that CA from forced rotation all stay valid until they expire, without limit on the number of them.

Reply:
My understanding is that we only want to trust 2 CAs - the current one and the previous one.  If more than one rotation is performed, the previous CA becomes untrusted for any client that refreshes.

Reply:
with the unbounded trust chain for manual rotation is not supported, disregard my comment.